src/main/scala/is/hail/utils/UInt.scala

```diff
@@ -46,25 +50,57 @@ class UInt private(val i: Int) extends AnyVal {
     if (l == r) 0 else if (l > r) 1 else -1
   }
 
-  def +(right: UInt): UInt = UInt(toLong + right.toLong)
-
-  def -(right: UInt): UInt = UInt(toLong - right.toLong)
-
-  def *(right: UInt): UInt = UInt(toLong * right.toLong)
-
-  def /(right: UInt): UInt = UInt(toLong / right.toLong)
-
-  def <=(right: UInt): Boolean = toLong <= right.toLong
-
-  def >=(right: UInt): Boolean = toLong >= right.toLong
-
-  def <(right: UInt): Boolean = toLong < right.toLong
-
-  def >(right: UInt): Boolean = toLong > right.toLong
-
-  def ==(right: UInt): Boolean = toLong == right.toLong
-
-  def !=(right: UInt): Boolean = toLong != right.toLong
+  def intRep: Int = i
+
+  def +(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong + right.toLong)
+  def +(right: Int): UInt = UInt(toLong + right)
+  def +(right: Double): Double = toDouble + right
+  def +(right: Long): Long = toLong + right
+
+  def -(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong - right.toLong)
+  def -(right: Int): UInt = UInt(toLong - right)
+  def -(right: Double): Double = toDouble - right
+  def -(right: Long): Long = toLong - right
+
+  def *(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong * right.toLong)
+  def *(right: Int): UInt = UInt(toLong * right)
+  def *(right: Double): Double = toDouble * right
+  def *(right: Long): Long = toLong * right
+
+  def /(right: UInt)(implicit uct: ClassTag[UInt]): UInt = UInt(toLong / right.toLong)
+  def /(right: Int): UInt = UInt(toLong / right)
+  def /(right: Double): Double = toDouble / right
+  def /(right: Long): Long = toLong / right
+
+  def <=(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong <= right.toLong
+  def <=(right: Int): Boolean = toLong <= right
+  def <=(right: Double): Boolean = toDouble <= right
+  def <=(right: Long): Boolean = toLong <= right
+
+  def >=(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong >= right.toLong
+  def >=(right: Int): Boolean = toLong >= right
+  def >=(right: Double): Boolean = toDouble >= right
+  def >=(right: Long): Boolean = toLong >= right
+
+  def <(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong < right.toLong
+  def <(right: Int): Boolean = toLong < right
+  def <(right: Double): Boolean = toDouble < right
+  def <(right: Long): Boolean = toLong < right
+
+  def >(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong > right.toLong
+  def >(right: Int): Boolean = toLong > right
+  def >(right: Double): Boolean = toDouble > right
+  def >(right: Long): Boolean = toLong > right
+
+  def ==(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong == right.toLong
+  def ==(right: Int): Boolean = toLong == right.toLong
+  def ==(right: Double): Boolean = toDouble == right
+  def ==(right: Long): Boolean = toLong == right
+
+  def !=(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = toLong != right.toLong
```

Comment:
I compared the behavior with and without the implicit class tag. If I don't have a second definition for the right-hand argument being an Int, the argument is implicitly converted to a Long and then the Long definition for `>` is called.

**With UInt implicit class:**
```scala
def >(right: UInt)(implicit uct: ClassTag[UInt]): Boolean = {
    println(s"right=$right type=${right.getClass()}")
    toLong > right.toLong
  }
  def >(right: Int): Boolean = {
    println(s"right=$right type=${right.getClass()}")
    toLong > right.toLong
  }
  def >(right: Double): Boolean = {
    println(s"right=$right type=${right.getClass()}")
    toDouble > right
  }
  def >(right: Long): Boolean = {
    println(s"right=$right type=${right.getClass()}")
    toLong > right
  }

>>> UInt(4294967295L) > UInt(0)
right=0 type=class is.hail.utils.UInt
>>> UInt(4294967295L) > -1
right=-1 type=int
>>> UInt(5) > 4.5
right=4.5 type=double
>>> UInt(5) > -1L
right=-1 type=long
>>> UInt(8) > 7
right=7 type=int
>>> UInt(8) > UInt(6)
right=6 type=class is.hail.utils.UInt
```

**Without UInt implicit class**:
```scala
  def >(right: UInt): Boolean = {
    println(s"right=$right type=${right.getClass()}")
    toLong > right.toLong
  }
  def >(right: Double): Boolean = {
    println(s"right=$right type=${right.getClass()}")
    toDouble > right
  }
  def >(right: Long): Boolean = {
    println(s"right=$right type=${right.getClass()}")
    toLong > right
  }

>>> UInt(4294967295L) > UInt(0)
right=0 type=class is.hail.utils.UInt
>>> UInt(4294967295L) > -1
right=-1 type=long
>>> UInt(5) > 4.5
right=4.5 type=double
>>> UInt(5) > -1L
right=-1 type=long
>>> UInt(8) > 7
right=7 type=long
>>> UInt(8) > UInt(6)
right=6 type=class is.hail.utils.UInt
```